File: packages/toi/src/sign.ts
/**
 * Ed25519 signing and verification over the RFC 8785 canonical form.
 *
 * The signed payload is always `canonicalize(document without $signature)`
 * encoded as UTF-8. Because canonicalization is order- and formatting-
 * independent, a signature survives reformatting, key reordering, and
 * round-tripping through any conformant parser.
 *
 * Crypto is provided by `@noble/ed25519` (audited, dependency-light). Its
 * synchronous API needs a SHA-512 implementation wired in once at module load.
 */
import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha512";
import { canonicalizeToBytes } from "./canonicalize.js";
import { bytesToBase64Url, base64UrlToBytes } from "./b64url.js";
import { parseToi } from "./parse.js";
import { ToiSignatureError } from "./errors.js";
import type { ToiDocument, ToiSignature } from "./types.js";

// Wire the synchronous SHA-512 that @noble/ed25519 v2 needs for sync sign/verify.
ed.etc.sha512Sync = (...messages: Uint8Array[]): Uint8Array =>
  sha512(ed.etc.concatBytes(...messages));

/** An Ed25519 key pair. Keys are raw 32-byte seeds / public points. */
export interface ToiKeyPair {
  /** 32-byte Ed25519 private seed. Keep secret; never write it into a `.toi` file. */
  privateKey: Uint8Array;
  /** 32-byte Ed25519 public key. */
  publicKey: Uint8Array;
  /** The public key as base64url — the form stored in `$signature.public_key`. */
  publicKeyBase64Url: string;
}

/** Generate a fresh Ed25519 key pair. */
export function generateKeyPair(): ToiKeyPair {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = ed.getPublicKey(privateKey);
  return { privateKey, publicKey, publicKeyBase64Url: bytesToBase64Url(publicKey) };
}

/** The exact bytes that get signed: the canonical form with `$signature` removed. */
export function signingPayload(doc: ToiDocument): Uint8Array {
  return canonicalizeToBytes(withoutSignature(doc));
}

/** `true` when `doc` carries a `$signature` envelope (not a validity claim). */
export function isSigned(doc: unknown): boolean {
  return isPlainObject(doc) && isPlainObject((doc as Record<string, unknown>)["$signature"]);
}

/**
 * Sign a document, returning a copy with a populated `$signature` field. The
 * input is validated first, so signing a malformed document throws.
 *
 * @throws {ToiValidationError} if `doc` is not a valid `.toi` document.
 * @throws {ToiSignatureError} if the cryptographic operation fails.
 */
export function signToi(doc: ToiDocument, privateKey: Uint8Array): ToiDocument {
  const validated = parseToi(doc);
  const unsigned = withoutSignature(validated);
  const payload = canonicalizeToBytes(unsigned);
  let publicKey: Uint8Array;
  let signature: Uint8Array;
  try {
    publicKey = ed.getPublicKey(privateKey);
    signature = ed.sign(payload, privateKey);
  } catch (err) {
    throw new ToiSignatureError("Failed to sign .toi document", { cause: err });
  }
  const $signature: ToiSignature = {
    alg: "ed25519",
    public_key: bytesToBase64Url(publicKey),
    value: bytesToBase64Url(signature),
  };
  return { ...unsigned, $signature } as ToiDocument;
}

/**
 * Verify a document's embedded `$signature` against its canonical payload.
 * Returns `false` for a missing, malformed, or non-matching signature.
 *
 * @throws {ToiSignatureError} only if the verification cannot be attempted
 *   (e.g. the encoded public key / value are not valid base64url).
 */
export function verifyToi(doc: ToiDocument): boolean {
  const raw = (doc as Record<string, unknown>)["$signature"];
  if (!isPlainObject(raw)) return false;
  if (raw["alg"] !== "ed25519") return false;
  const publicKeyB64 = raw["public_key"];
  const valueB64 = raw["value"];
  if (typeof publicKeyB64 !== "string" || typeof valueB64 !== "string") return false;

  let publicKey: Uint8Array;
  let signature: Uint8Array;
  try {
    publicKey = base64UrlToBytes(publicKeyB64);
    signature = base64UrlToBytes(valueB64);
  } catch (err) {
    throw new ToiSignatureError("Malformed signature encoding", { cause: err });
  }

  const payload = canonicalizeToBytes(withoutSignature(doc));
  try {
    return ed.verify(signature, payload, publicKey);
  } catch {
    // Wrong-length key/signature bytes are simply not verifiable.
    return false;
  }
}

function withoutSignature(doc: ToiDocument): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  const source = doc as Record<string, unknown>;
  for (const key of Object.keys(source)) {
    if (key !== "$signature") out[key] = source[key];
  }
  return out;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: src/nlt_toi/sign.py
"""Ed25519 signing and verification over the RFC 8785 canonical form.

The signed payload is always ``canonicalize(document without $signature)`` encoded
as UTF-8. Because canonicalization is order- and formatting-independent, a
signature survives reformatting, key reordering, and round-tripping through any
conformant parser — and interoperates with the reference TypeScript
implementation (a token signed by one verifies in the other).

Crypto is provided by the audited ``cryptography`` library; keys are raw 32-byte
Ed25519 seeds / public points, matching the ``$signature`` envelope (SPEC §11).
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .b64url import base64url_to_bytes, bytes_to_base64url
from .canonicalize import canonicalize_to_bytes
from .errors import ToiSignatureError
from .parse import ToiDocument, parse_toi

#: SPEC §11.1: signature fields are unpadded base64url — no ``=`` padding, no whitespace.
_UNPADDED_BASE64URL = re.compile(r"^[A-Za-z0-9_-]+$")

_RAW = serialization.Encoding.Raw
_RAW_PRIVATE = serialization.PrivateFormat.Raw
_RAW_PUBLIC = serialization.PublicFormat.Raw
_NO_ENCRYPTION = serialization.NoEncryption()


@dataclass(frozen=True)
class ToiKeyPair:
    """An Ed25519 key pair. Keys are raw 32-byte seeds / public points."""

    #: 32-byte Ed25519 private seed. Keep secret; never write it into a ``.toi`` file.
    private_key: bytes
    #: 32-byte Ed25519 public key.
    public_key: bytes
    #: The public key as base64url — the form stored in ``$signature.public_key``.
    public_key_base64url: str


def generate_key_pair() -> ToiKeyPair:
    """Generate a fresh Ed25519 key pair."""
    private = Ed25519PrivateKey.generate()
    private_bytes = private.private_bytes(_RAW, _RAW_PRIVATE, _NO_ENCRYPTION)
    public_bytes = private.public_key().public_bytes(_RAW, _RAW_PUBLIC)
    return ToiKeyPair(private_bytes, public_bytes, bytes_to_base64url(public_bytes))


def signing_payload(doc: ToiDocument) -> bytes:
    """The exact bytes that get signed: the canonical form with ``$signature`` removed."""
    return canonicalize_to_bytes(_without_signature(doc))


def is_signed(doc: Any) -> bool:
    """``True`` when *doc* carries a ``$signature`` envelope (not a validity claim)."""
    return _is_plain_object(doc) and _is_plain_object(doc.get("$signature"))


def sign_toi(doc: ToiDocument, private_key: bytes) -> ToiDocument:
    """Sign a document, returning a copy with a populated ``$signature`` field.

    The input is validated first, so signing a malformed document raises.

    Raises:
        ToiValidationError: if *doc* is not a valid ``.toi`` document.
        ToiCanonicalizationError: if *doc* holds values not representable as
            canonical JSON.
        ToiSignatureError: if the cryptographic operation fails.
    """
    validated = parse_toi(doc)
    unsigned = _without_signature(validated)
    payload = canonicalize_to_bytes(unsigned)
    try:
        private = Ed25519PrivateKey.from_private_bytes(private_key)
        signature = private.sign(payload)
        public_bytes = private.public_key().public_bytes(_RAW, _RAW_PUBLIC)
    except Exception as err:  # noqa: BLE001 — wrapped into the library's taxonomy
        raise ToiSignatureError("Failed to sign .toi document") from err
    unsigned["$signature"] = {
        "alg": "ed25519",
        "public_key": bytes_to_base64url(public_bytes),
        "value": bytes_to_base64url(signature),
    }
    return unsigned


def verify_toi(doc: ToiDocument) -> bool:
    """Verify a document's embedded ``$signature`` against its canonical payload.

    Fully defensive: returns ``False`` for a missing, malformed, undecodable, or
    non-matching signature, and never raises.
    """
    if not _is_plain_object(doc):
        return False
    raw = doc.get("$signature")
    if not _is_plain_object(raw):
        return False
    if raw.get("alg") != "ed25519":
        return False
    public_key_b64 = raw.get("public_key")
    value_b64 = raw.get("value")
    if not isinstance(public_key_b64, str) or not isinstance(value_b64, str):
        return False
    # SPEC §11.1: reject padded / whitespaced encodings rather than normalizing them.
    if not _UNPADDED_BASE64URL.match(public_key_b64) or not _UNPADDED_BASE64URL.match(value_b64):
        return False
    try:
        public_bytes = base64url_to_bytes(public_key_b64)
        signature = base64url_to_bytes(value_b64)
    except ValueError:
        return False
    try:
        payload = canonicalize_to_bytes(_without_signature(doc))
        Ed25519PublicKey.from_public_bytes(public_bytes).verify(signature, payload)
        return True
    except Exception:  # noqa: BLE001 — un-canonicalizable content / bad key bytes / mismatch: not verifiable
        return False


def _without_signature(doc: ToiDocument) -> ToiDocument:
    return {key: value for key, value in doc.items() if key != "$signature"}


def _is_plain_object(value: Any) -> bool:
    return isinstance(value, dict)
